### PR TITLE
Update instrumentation test configuration

### PR DESCRIPTION
**Pull request by f2rf2r**
- Add a few more common dependencies for androidTest
- Move mockServerRule from unit test to androidTest
- Add setup for overriding restApi and userSettings during test
- Create base classes for PresenterTest and UiTest
- Update example to involve using retrofit (and mocking it during instrumentation test)

### CODE REVIEW

File: app/src/main/java/io/intrepid/skeleton/screens/example2/Example2Fragment.java [hunk 1]
```diff
@@ -25,7 +25,7 @@ protected int getLayoutResourceId() {
     }
 
     @Override
-    public void showText(String text) {
-        textView.setText(text);
+    public void showIpAddress(String ip) {
+        textView.setText("Your Ip address is " + ip);
```
Comment (gdaniels) [inline app/src/main/java/io/intrepid/skeleton/screens/example2/Example2Fragment.java]: Despite this just being an example we should probably externalize this string - or at least put in a comment that describes why not.

File: app/src/androidTest/java/io/intrepid/skeleton/testutils/BaseUiTest.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.After;
+import org.junit.Rule;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+import io.intrepid.skeleton.InstrumentationTestApplication;
+
+public class BaseUiTest {
+    @Rule
+    public MockitoRule mockitoRule = MockitoJUnit.rule();
```
Comment (gdaniels) [inline app/src/androidTest/java/io/intrepid/skeleton/testutils/BaseUiTest.java:12]: If we're putting in the capability for instrumentation test mocks, should we give an example like we do on the unit test side?

**Comment by gdaniels** [comment]
Couple of small comments but basically 👍  from me.  🔧 + 🚢

File: app/src/androidTest/java/io/intrepid/skeleton/testutils/BaseUiTest.java (hunk 2 above)
Comment (f2rf2r) [inline app/src/androidTest/java/io/intrepid/skeleton/testutils/BaseUiTest.java]: Yep, I will add a unit test for Example2Presenter

Comment (f2rf2r) [reply app/src/androidTest/java/io/intrepid/skeleton/testutils/BaseUiTest.java:12]: I will add an example with mock UserSettings

File: app/src/main/java/io/intrepid/skeleton/base/PresenterConfiguration.java [hunk 3]
```diff
@@ -6,11 +6,14 @@
...
 public class PresenterConfiguration {
     @NonNull
-    protected final Scheduler ioScheduler;
+    private final Scheduler ioScheduler;
     @NonNull
-    protected final Scheduler uiScheduler;
+    private final Scheduler uiScheduler;
     @NonNull
     private final UserSettings userSettings;
     @NonNull
```
Comment (shadow7) [inline app/src/main/java/io/intrepid/skeleton/base/PresenterConfiguration.java:20]: Do we want to add `CrashReporter` and `AnalyticsTracker` as injected dependencies to the presenter config?

Comment (f2rf2r) [reply app/src/main/java/io/intrepid/skeleton/base/PresenterConfiguration.java:20]: I can add the crashReporter. Analytics is currently not part of the skeleton app

**Review by shadow7 — APPROVED** [review]

File: app/src/androidTest/java/io/intrepid/skeleton/InstrumentationTestApplication.java [hunk 4]
```diff
@@ -2,22 +2,45 @@
...
-                // these can be replaced with mock version
-                SharePreferencesManager.getInstance(this),
-                RetrofitClient.getApi()
-        );
+                userSettingsOverride != null ? userSettingsOverride : SharePreferencesManager.getInstance(this),
+                restApiOverride != null ? restApiOverride : RetrofitClient.getApi(),
+                Mockito.mock(CrashReporter.class));
+    }
+
+    public static void overrideResApi(RestApi restApi) {
```
Comment (prfarlow1) [inline app/src/androidTest/java/io/intrepid/skeleton/InstrumentationTestApplication.java]: Missing a T here

**Pull request merged**